File: Find-Lucky-Integer-in-an-Array/FLIiaA.rs
pub fn find_lucky(arr: Vec<i32>) -> i32 {
    let mut cache = [0; 501];
    for n in arr {
        cache[n as usize] += 1;
    }
    for i in (1..=500).rev() {
        if cache[i] == i {
            return i as i32;
        }
    }

    return -1;
}

pub fn find_lucky2(arr: Vec<i32>) -> i32 {
    let mut cache = [0; 501];
    arr.iter().fold(&mut cache, |cache, &n| {
        cache[n as usize] += 1;
        cache
    });
    (1..=500)
        .rev()
        .find(|&x| cache[x as usize] == x)
        .unwrap_or(-1)
}

fn main() {
    assert_eq!(2, find_lucky2(vec![2, 2, 3, 4]));
    assert_eq!(3, find_lucky2(vec![1, 2, 2, 3, 3, 3]));
    assert_eq!(-1, find_lucky2(vec![2, 2, 2, 3, 3]));
    assert_eq!(-1, find_lucky2(vec![5]));
    assert_eq!(7, find_lucky2(vec![7, 7, 7, 7, 7, 7, 7]));
}

File: Find-Lucky-Integer-in-an-Array/FLIiaA.c
#include <assert.h>
#include <stdio.h>

int findLucky(int* arr, int arrSize) {
  int cache[501] = {};
  for (int i = 0; i < arrSize; i++) {
    cache[arr[i]] += 1;
  };

  for (int i = 500; i > 0; i--) {
    if (cache[i] == i) {
      return i;
    }
  }

  return -1;
}

int main() {
  int case0[4] = {2, 2, 3, 4};
  assert(2 == findLucky(case0, 4));
}
